[SABRA-2456] Add V2 Facets and Searchabilities API Support

aandrukhovich opened the pull request:
## Summary

This PR adds support for the v2 versions of the facets and searchabilities APIs, which include additional fields and capabilities not available in v1.

### Key Changes

**New Model Classes:**
- `FacetConfigurationV2` - v2 facet configuration with `path_in_metadata` field support
- `SearchabilityV2` - Searchability configuration model

**New Request Classes:**
- `FacetConfigurationV2Request` - Single facet v2 operations
- `FacetConfigurationsV2Request` - Bulk facet v2 operations
- `SearchabilityV2Request` - Single searchability operations
- `SearchabilitiesV2Request` - Bulk searchability PATCH operations
- `SearchabilitiesV2GetRequest` - Searchability list operations with filtering/pagination
- `SearchabilitiesV2DeleteRequest` - Bulk searchability deletion

**New API Methods in `ConstructorIO`:**

| Endpoint | Method |
|----------|--------|
| `GET /v2/facets` | `retrieveFacetConfigurationsV2()` |
| `GET /v2/facets/{name}` | `retrieveFacetConfigurationV2()` |
| `POST /v2/facets` | `createFacetConfigurationV2()` |
| `PUT /v2/facets/{name}` | `replaceFacetConfigurationV2()` |
| `PATCH /v2/facets/{name}` | `updateFacetConfigurationV2()` |
| `PATCH /v2/facets` | `updateFacetConfigurationsV2()` |
| `PUT /v2/facets` | `replaceFacetConfigurationsV2()` |
| `DELETE /v2/facets/{name}` | `deleteFacetConfigurationV2()` |
| `GET /v2/searchabilities` | `retrieveSearchabilitiesV2()` |
| `GET /v2/searchabilities/{name}` | `retrieveSearchabilityV2()` |
| `PATCH /v2/searchabilities/{name}` | `createOrUpdateSearchabilityV2()` |
| `PATCH /v2/searchabilities` | `createOrUpdateSearchabilitiesV2()` |
| `DELETE /v2/searchabilities/{name}` | `deleteSearchabilityV2()` |
| `DELETE /v2/searchabilities` | `deleteSearchabilitiesV2()` |

### v2 API Differences from v1

- **Facets v2:** Requires `path_in_metadata` field which specifies where in item metadata the facet data is located
- **Searchabilities v2:** New endpoints for managing searchability configurations with support for `skip_rebuild` parameter

## Test Plan

- [ ] Unit tests for all model and request classes
- [ ] Integration tests for all v2 API methods
- [ ] Verify facet CRUD operations work correctly
- [ ] Verify searchability CRUD operations work correctly
- [ ] Verify bulk operations work correctly
- [ ] Verify pagination and filtering work for list operations

aandrukhovich committed:
Add v2 versions of facets and searchabilities

aandrukhovich committed:
fix tests

aandrukhovich committed:
update api key: use special key for v2 facets and searchabilities test

aandrukhovich committed:
don't pass name

aandrukhovich committed:
return name

aandrukhovich committed:
fix expected 401 message

aandrukhovich committed:
address comments

aandrukhovich committed:
formatting

aandrukhovich committed:
address comments

aandrukhovich committed:
address more comments

aandrukhovich committed:
remove

aandrukhovich committed:
remove

aandrukhovich committed:
remove

aandrukhovich committed:
empty line

esezen committed:
Better section logic

esezen committed:
Handle edge cases

esezen committed:
Update tests

esezen committed:
Merge branch 'master' into SABRA-2456/cio-java

esezen committed:
Type validation

esezen committed:
Lint

esezen reviewed (dismissed):
We need one more review

HHHindawy commented on constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:
```diff
@@ -3489,4 +3520,773 @@ public SortOptionsResponse retrieveSortOptions(String sortBy) throws Constructor
+     * @throws IllegalArgumentException if request is null
+     * @throws ConstructorException if the request fails
+     */
+    public String replaceFacetConfigurationsV2(
```
Should we also run `validateFacetConfigurationV2Type` here on every facet configuration?

esezen replied:
Good catch, added it. `replaceFacetConfigurationsV2` now runs `validateFacetConfigurationV2Type` on each facet configuration, matching the single replace/create endpoints (bulk replace is a full replace, so `type` is required). Left the PATCH update methods as-is since `type` is optional for partial updates. Added a test covering the invalid-type case. (commit `9d3eb34e`)

HHHindawy reviewed (dismissed):
Approved, with a small suggestion 🙏 

Also, @esezen looks like Quizzes tests are failing, but it's unrelated to this PR.

esezen committed:
Merge branch 'master' into SABRA-2456/cio-java

esezen committed:
Validate facet type in bulk replace V2

Run validateFacetConfigurationV2Type on each facet configuration in
replaceFacetConfigurationsV2 for parity with the single-replace and
create endpoints, since bulk replace is a full replace where type is
required.

HHHindawy reviewed (approved):
LGTM!
